src/lib/device/device_manager.cpp

```diff
@@ -462,348 +48,349 @@ int DeviceManager::get_open_flags(const io_flag oflags) {
-
-#ifndef NO_CHECKSUM
-    dm_derived.info->checksum = crc16_t10dif(
-        hs_init_crc_16, reinterpret_cast< const unsigned char* >(chunk_memory + dm_info::s_dm_payload_offset),
-        dm_derived.info_size - dm_info::s_dm_payload_offset);
-#endif
-
-    for (uint32_t i{0}; i < dm_derived.pdev_hdr->num_phys_devs; ++i) {
-        auto* pdev = get_pdev(i);
-        pdev->write_dm_chunk(gen_count, reinterpret_cast< const char* >(chunk_memory), dm_derived.info_size);
+DeviceManager::DeviceManager(const std::vector< dev_info >& devs, vdev_create_cb_t vdev_create_cb) :
+        m_dev_infos{devs}, m_vdev_create_cb{std::move(vdev_create_cb)} {
+    bool found_hdd_dev{false};
+    for (const auto& dev_info : devs) {
+        if (is_hdd(dev_info.dev_name)) {
+            HomeStoreStaticConfig::instance().hdd_drive_present = true;
+            found_hdd_dev = true;
+            LOGINFO("HDD device found: {}");
+            break;
+        }
     }
 
-    HS_DBG_ASSERT_EQ(dm_derived.vdev_hdr->get_magic(), MAGIC);
-    HS_DBG_ASSERT_EQ(dm_derived.chunk_hdr->get_magic(), MAGIC);
-    HS_DBG_ASSERT_EQ(dm_derived.pdev_hdr->get_magic(), MAGIC);
-}
-
-PhysicalDevChunk* DeviceManager::alloc_chunk(PhysicalDev* pdev, const uint32_t vdev_id, const uint64_t req_size,
-                                             const uint32_t primary_id) {
-    std::lock_guard< decltype(m_dev_mutex) > lock{m_dev_mutex};
-
-    HS_DBG_ASSERT_EQ(req_size % pdev->page_size(), 0);
-    PhysicalDevChunk* chunk = pdev->find_free_chunk(req_size);
-    if (chunk == nullptr) {
-        LOGERROR("No space available for chunk size {} in pdev id {} ", in_bytes(req_size), pdev->dev_id());
-        return nullptr;
+    if (found_hdd_dev) {
+        if ((HS_STATIC_CONFIG(input.data_open_flags) == io_flag::DIRECT_IO) &&
+            !HS_DYNAMIC_CONFIG(device->direct_io_mode)) {
+            // override direct i/o for HDD's
+            LOGINFO("Overridding HDD open flags from DIRECT_IO to BUFFERED_IO");
+            m_hdd_open_flags = determine_open_flags(io_flag::BUFFERED_IO);
+        } else {
+            m_hdd_open_flags = determine_open_flags(HS_STATIC_CONFIG(input.data_open_flags));
+        }
     }
-    HS_DBG_ASSERT_GE(chunk->size(), req_size);
-    chunk->set_vdev_id(vdev_id); // Set the chunk as busy or engaged to a vdev
-    chunk->set_primary_chunk_id(primary_id);
-    chunk->update_end_of_chunk(req_size);
-
-    // if it is more then the requested size than trim chunk from the end.
-    if (chunk->size() > req_size) {
-        // There is some left over space, create a new chunk and insert it after current chunk
-        create_new_chunk(pdev, chunk->start_offset() + req_size, chunk->size() - req_size, chunk);
-        chunk->set_size(req_size);
+    m_ssd_open_flags = determine_open_flags(HS_STATIC_CONFIG(input.fast_open_flags));
+
+    // Read from all the devices and check if there is a valid superblock present in those devices.
+    m_first_time_boot = true;
```

Comment:
nit: can be removed if initialized in class;

Reply:
I initialized in class, but better here to set this explicitly in constructor as well.